Use NaN as default resolution for TimeSeries objects

t-b opened the pull request:
The NWB schema documentation says

Smallest meaningful difference between values in data, stored in the specified by unit.
Comment: E.g., the change in value of the least significant bit, or a
larger number if signal noise is known to be present. If unknown, use NaN.

So using NaN instead of zero is a more natural choice as a resolution of zero
means that you have perfect experimental conditions which are in general
rarely available.

t-b committed:
Use NaN as default resolution for TimeSeries objects

The NWB schema documentation says

Smallest meaningful difference between values in data, stored in the specified by unit.
Comment: E.g., the change in value of the least significant bit, or a
larger number if signal noise is known to be present. If unknown, use NaN.

So using NaN instead of zero is a more natural choice as a resolution of zero
means that you have perfect experimental conditions which are in general
rarely available.

pep8speaks commented:
Hello @t-b! Thanks for updating the PR.











Cheers ! There are no PEP8 issues in this Pull Request. :beers: 

##### Comment last updated on April 16, 2018 at 22:26 Hours UTC

t-b commented:
Okay I'm not understanding the tests well enough to get them fixed. Somehow the resolution is still stuck to 0 at some place.

t-b committed:
TestMapNWBContainer: Add assertDictEqual from Python 3 (debian stretch)

We want to adapt the behaviour in a later commit.

t-b committed:
TestMapNWBContainer: Fix assertDictEqual for nan values in the dictionary

Having dictionaries with NaN entries results in assertDictEqual saying
that the two dictionaries are not equal as NaN == NaN is False.

We workaround that issue by replacing known entries if they are NaN with
a constant.

t-b commented:
@ajtritt Tests work :)

ajtritt commented on tests/integration/ui_write/base.py:
```diff
@@ -45,6 +47,22 @@ def container(self):
+
+        if d1 != d2 and len('\n'.join(difflib.context_diff(pprint.pformat(d1).splitlines(),
+            pprint.pformat(d2).splitlines()))) > 0:
+            standardMsg = '%s != %s' % _common_shorten_repr(d1, d2)
```
Looks like you missed the function `_common_shorten_repr`  when copying over the assert method.

t-b replied:
@ajtritt Thanks, I'll fix that.

t-b replied:
@ajtritt Done.

ajtritt reviewed (changes requested):
Fix missing function issue

t-b commented:
@ajtritt Fixed.

t-b commented:
@ajtritt Any hints why the circleci python 2.7 job does not pass?

ajtritt commented:
Looks like the equality comparison is breaking on unicode-ascii comparisons.

t-b commented:
@ajtritt Hmm... But the miniconda27 passes. I look into that when I get a change.

t-b commented:
I can reproduce the failure locally with python 2.7.

t-b commented:
@ajtritt I finally found a not too hacky way...

t-b committed:
TimeSeries: Rename rate_unit variable and use it correctly

Since 575da105 (moved base TimeSeries class into submodule core,
2017-01-30) the unit of the rate is defined as "Seconds". But in
TimeSeriesMap it is used as starting time unit.

Rename the class variable therefore from rate_unit to
starting_time_unit.

A user visible change is that we now fill starting_time_unit only when a
starting_time is given.

t-b commented:
@ajtritt Regarding the review: I (think) I adressed your comments. Only github seems not to agree.

t-b committed:
Merge branch 'dev' into proper-resolution

t-b commented:
@oruebel @ajtritt With the approaching release, is that good to go?

ajtritt commented:
@t-b #610 addressed this. I changed resolution to default to -1.0. The logic being, since resolution is a difference, it should never be positive, so it's just as informative in this context as NaN. Using -1.0 has the added benefit of being a floating point number that behaves as one would expect a floating point number, namely `-1.0 == -1.0` is *true*, while `NaN == NaN` is *false* 

Does this meet your needs?

t-b commented:
@ajtritt Thanks for the update. From a purist technical POV I still prefer `Nan` but a negative resolution does the trick as well.
